### PR TITLE
Raise a warning when user-applied `dynamic_one_shot` is ignored

### PR DESCRIPTION
**Pull request by astralcai**
Fixes https://github.com/PennyLaneAI/pennylane/issues/6551
[sc-77877]

### CODE REVIEW

File: pennylane/transforms/dynamic_one_shot.py [hunk 1]
```diff
@@ -58,40 +58,42 @@ def null_postprocessing(results):
...
     **Example**
 
-    Consider the following circuit:
+    Most devices that support mid-circuit measurements will include this transform in its
+    preprocessing automatically when applicable. When this is the case, any user-applied
+    ``dynamic_one_shot`` transforms will be ignored. The recommended way to use dynamic one
+    shot is to specify ``mcm_method="one-shot"`` in the ``qml.qnode`` decorator.
 
-    .. code-block:: python
+    ..code-block:: python
```
Comment (astralcai) [inline pennylane/transforms/dynamic_one_shot.py:85]:
```suggestion
    .. code-block:: python
```

File: pennylane/transforms/dynamic_one_shot.py [hunk 2]
```diff
@@ -58,40 +58,42 @@ def null_postprocessing(results):
 def dynamic_one_shot(tape: QuantumScript, **kwargs) -> tuple[QuantumScriptBatch, PostprocessingFn]:
     """Transform a QNode to into several one-shot tapes to support dynamic circuit execution.
 
+    This transform enables the ``"one-shot"`` mcm method. The ``"one-shot"`` method prompts the
+    device to perform a series of one-shot executions, where in each execution, the ``qml.measure``
+    operation dynamically measures the wire and collapse the state vector stochastically.
```
Comment (isaacdevlugt) [inline pennylane/transforms/dynamic_one_shot.py:63]:
```suggestion
    operation dynamically measures the wire and collapses the state vector stochastically.
```

File: pennylane/transforms/dynamic_one_shot.py [hunk 3]
```diff
@@ -58,40 +58,42 @@ def null_postprocessing(results):
...
+    This transform contracts with ``qml.defer_measurement``, which instead introduces an extra
```
Comment (isaacdevlugt) [inline pennylane/transforms/dynamic_one_shot.py:64]:
```suggestion
    This is in contrast with ``qml.defer_measurement``, which instead introduces an extra
```

File: pennylane/transforms/dynamic_one_shot.py [hunk 4]
```diff
@@ -58,40 +58,42 @@ def null_postprocessing(results):
...
+    wire for each mid-circuit measurement. The ``"one-shot"`` method is favourable in the few-shots
+    several-mid-circuit-measurements limit, whereas ``qml.defer_measurements`` is favourable in
```
Comment (isaacdevlugt) [inline pennylane/transforms/dynamic_one_shot.py:66]:
```suggestion
    wire for each mid-circuit measurement. The ``"one-shot"`` method is favourable in the few-shots
    and several-mid-circuit-measurements limit, whereas ``qml.defer_measurements`` is favourable in
```

File: pennylane/workflow/_setup_transform_program.py [hunk 5]
```diff
@@ -46,10 +48,20 @@ def _prune_dynamic_transform(outer_transform, inner_transform):
...
+    inner_contains_one_shot = inner_transform.prune_dynamic_transform(type_to_keep)
+    if inner_contains_one_shot:
         type_to_keep = 0
+    original_len = len(outer_transform)
     outer_transform.prune_dynamic_transform(type_to_keep)
+    outer_contained_one_shot = len(outer_transform) < original_len
+    if inner_contains_one_shot and outer_contained_one_shot:
+        warnings.warn(
+            "A dynamic_one_shot transform already exists in the preprocessing program of the "
+            "device, therefore, the dynamic_one_shot applied on the qnode as a decorator is "
```
Comment (isaacdevlugt) [inline pennylane/workflow/_setup_transform_program.py:60]:
```suggestion
            "device. Therefore, the dynamic_one_shot applied on the qnode as a decorator is "
```

**Review by isaacdevlugt — APPROVED** [review]
Thanks @astralcai !

File: pennylane/transforms/dynamic_one_shot.py [hunk 6]
```diff
@@ -58,40 +58,42 @@ def null_postprocessing(results):
 def dynamic_one_shot(tape: QuantumScript, **kwargs) -> tuple[QuantumScriptBatch, PostprocessingFn]:
     """Transform a QNode to into several one-shot tapes to support dynamic circuit execution.
 
+    This transform enables the ``"one-shot"`` mcm method. The ``"one-shot"`` method prompts the
+    device to perform a series of one-shot executions, where in each execution, the ``qml.measure``
+    operation dynamically measures the wire and collapses the state vector stochastically.
```
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:63]: This is information specific to state-vector simulators. `one-shot` only transforms the circuit, and it's up to the device to figure out how to support it. A device in hardware could use one-shot, in which case `dynamic_one_shot`'s job would be to post-process the results correctly.

File: pennylane/workflow/_setup_transform_program.py [hunk 7]
```diff
@@ -46,10 +48,20 @@ def _prune_dynamic_transform(outer_transform, inner_transform):
...
+    original_len = len(outer_transform)
     outer_transform.prune_dynamic_transform(type_to_keep)
+    outer_contained_one_shot = len(outer_transform) < original_len
+    if inner_contains_one_shot and outer_contained_one_shot:
+        warnings.warn(
+            "A dynamic_one_shot transform already exists in the preprocessing program of the "
+            "device. Therefore, the dynamic_one_shot applied on the qnode as a decorator is "
+            "ignored. See https://docs.pennylane.ai/en/latest/code/api/pennylane.dynamic_one_shot.html "
+            "for more information on the recommended way to use dynamic one shot.",
+            UserWarning,
```
Comment (mudit2812) [inline pennylane/workflow/_setup_transform_program.py:63]:
```suggestion
            "A dynamic_one_shot transform already exists in the preprocessing program of the "
            "device. Therefore, the dynamic_one_shot applied on the qnode will be ignored. "
            "See https://docs.pennylane.ai/en/stable/code/api/pennylane.dynamic_one_shot.html "
            "for more information on the recommended way to use dynamic_one_shot.",
            UserWarning,
```

**Review by mudit2812 — COMMENTED** [review]
Looks good, few comments.

File: pennylane/transforms/dynamic_one_shot.py (hunk 6 above)
Comment (astralcai) [reply pennylane/transforms/dynamic_one_shot.py:63]:
```suggestion
    This transform prepares the circuit for the ``"one-shot"`` mcm method and post-process the
    results returned from the device. The ``"one-shot"`` method prompts the device to perform a 
    series of one-shot executions, where in each execution, the ``qml.measure`` operation dynamically 
    measures the wire and collapses the state vector stochastically.
```
How about now? @mudit2812

Comment (mudit2812) [reply pennylane/transforms/dynamic_one_shot.py:63]: state vectors shouldn't be mentioned at all. Maybe the following:
```suggestion
    This transform enables the ``"one-shot"`` mid-circuit measurement method. The ``"one-shot"`` method prompts the
    device to perform a series of one-shot executions, where in each execution, the ``qml.measure``
    operation applies a probabilistic mid-circuit measurement to the circuit.
```

**Pull request merged**